torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1167,60 +856,335 @@ def _get_ignored_params(
+        device is fixed to the current device.
+
+        NOTE: For now, the compute device is always a CUDA GPU device with its
+        explicit index.
+
+        Precondition: ``_check_single_device_module()``.
+        """
+        if not torch.cuda.is_available():
+            raise RuntimeError("FSDP does not support CPU only execution")
+        # If the module is on GPU already, then that GPU device has priority
+        # over the current device
+        param = next(self._get_orig_params(module, ignored_params), None)
+        if param is not None and param.device.type == "cuda":
+            compute_device = param.device
+        else:
+            compute_device = torch.device("cuda", torch.cuda.current_device())
+        if device_id is not None:
+            device = (
+                device_id
+                if isinstance(device_id, torch.device)
+                else torch.device(device_id)
+            )
+            if device == torch.device("cuda"):
+                warnings.warn(
+                    f"FSDP got the argument `device_id` {device_id} on rank "
+                    f"{self.rank}, which does not have an explicit index. "
+                    f"FSDP will use the current device {torch.cuda.current_device()}. "
+                    "If this is incorrect, please explicitly call `torch.cuda.set_device()` "
+                    "before FSDP initialization or pass in the explicit device "
+                    "index as the `device_id` argument."
+                )
+            elif device != compute_device:
+                raise AssertionError(
+                    "Inconsistent compute device and `device_id` on rank "
+                    f"{self.rank}: {compute_device} vs {device}"
+                )
+        return compute_device
 
-    @classmethod
-    def _check_wrapped(cls, begin_module, check_fn, err_fn):
-        for _, mod in begin_module.named_modules():
-            if not check_fn(mod):
-                raise ValueError(err_fn(mod))
+    def _materialize_module(
+        self,
+        module: nn.Module,
+        param_init_fn: Optional[Callable[[nn.Module], None]],
+    ) -> None:
+        """
+        Materializes the wrapped module ``module`` in place if needed: either
+        if the module has parameters that use meta device or are torchdistX
+        fake tensors.
+
+        This method uses ``param_init_fn`` to materialize the module if the
+        function is not ``None`` and falls back to default behavior otherwise.
+        For meta device, this calls ``reset_parameters()``, and for torchdistX
+        fake tensors, this calls ``deferred_init.materialize_module()``.
+
+        Precondition: ``self.compute_device`` is set via
+        ``_get_compute_device()``.
+        """
+        is_meta_module = any(p.is_meta for p in module.parameters())
+        is_torchdistX_deferred_init = (
+            not is_meta_module
+            and _TORCHDISTX_AVAIL
```

Comment:
I heard fake tensor has been upstreamed to PyTorch (with a different implementation). Is there any gap in the PyTorch fake tensor implementation, or are there any other reasons that FSDP still need to depend on torchdistx? (not calling for changes in this refactoring PR, just curious)

Reply:
We had a discussion with @ezyang and the conclusion was at that point, the API is not fully BC with torchdistX and long term, it likely will not be. There will be some collaborations involved in switching over to using fake tensors in core which we should ultimately do, but this needs to be scoped out